### ISSUE TITLE
Finding Git directory was broken by adding GIT_DIR support in the general case

### ISSUE THREAD

**Issue by mentalisttraceur** [eza-community/eza#115]
A problem introduced by how GIT_DIR support was added: makes Eza find the wrong Git directory in some cases.

For any given path argument given to Eza:

1. ✓ If `GIT_DIR` is set, and the path is *inside* that Git repo, great!

2. :x: If `GIT_DIR` is set, but the path is *outside* that Git repo, Eza **lies**: it prints the git status column but every line shows `--` as if there are no changes (the status check is done against the wrong repo; Eza never tried finding the Git repo for that path, since it found the one at GIT_DIR and was satisfied).

    1. :x: If the path is in *a different* Git repo, the ideal behavior would be to show the correct Git status.
    2. :x: If the path is *not in any* Git repo, the ideal behaviour would be to not even show the Git column.

3. ✓ If `GIT_DIR` is set to something that's not a gitdir, great! (`git2::Repository::open_from_env` returns a not-found error, so Eza falls through to trying to normal git repository lookup on the path.)

    1. ✓ If the path is in a Git repo, the normal lookup finds it and shows the correct Git status.
    2. ✓ If the path is not in a Git repo, the Git column is correctly omitted.

4. :x: If `GIT_DIR` is not set, `git2::Repository::open_from_env` will try searching *from the current directory* (instead of searching *from the given path*), so:

    1. ✓ if `eza` was executed inside a Git repo, and the path is inside the same repo, great (by coincidence)!

    2. :x: if `eza` was executed inside a Git repo, and the path is outside that repo, Eza **lies** (same as case 2).

    3. ✓ if `eza` was executed outside of a Git repo, great!

In particular, note that last :x: case is a *regression*: even if a user doesn't use `GIT_DIR`, if they just happen to be inside a Git repo when they run Eza, the output will now show wrong status information for paths outside that repo (potentially importantly wrong! `eza -l --git ../foo` is something a user might run before deciding to `rm -r ../foo` based on the status infomation!) I'm surprised there wasn't a test to catch this in the automated tests (or maybe there is, but it's in the Vagrant tests which no one seems to be able to run)?

**Comment by mentalisttraceur** [comment]
# Minimum Solution

To just fix the regression, instead of always trying `open_from_env` and then falling back to `discover`, we could pick one of those functions based on whether or not `GIT_DIR` is set (for example, with `env::var("GIT_DIR").is_ok()`).

**Comment by mentalisttraceur** [comment]
# Full Solution

We can treat `GIT_DIR` as an **additional** repo, not as a replacement for any of the normal searches. So if `GIT_DIR` is defined, we do one `open_from_env` call, and put that as the first item in the `repos: Vec<GitRepo>` of the `GitCache` struct (that way it gets checked first, but if a path isn't in it, we fall through to the normal Git repo search).

(Every other approach I thought of led to dead-ends or complications. I had a longer write-up of what's tricky about GIT_DIR for Eza, considering multiple possible path arguments at once and all that, but that got deleted and I don't feel like re-writing it right now - feel free to ask questions, etc.)

**Comment by cafkafk** [comment]
Wow, thanks for the write-up (annoying that some of it was deleted :c), it's very comprehensive!

I noticed you had a [fork](https://github.com/mentalisttraceur/eza), will you be submitting a PR for this?

I'll look more into this later this week when I have some hours for bug-crunching on eza.

**Comment by mentalisttraceur** [comment]
> will you be submitting a PR for this?

Not anymore. (because I just did :smiley: :tada: )

**Comment by iamkroot** [comment]
> For any given path argument given to Eza:
> 
>     1. ✓ If `GIT_DIR` is set, and the path is _inside_ that Git repo, great!
> 
>     2. ❌ If `GIT_DIR` is set, but the path is _outside_ that Git repo, Eza **lies**: it prints the git status column but every line shows `--` as if there are no changes (the status check is done against the wrong repo; Eza never tried finding the Git repo for that path, since it found the one at GIT_DIR and was satisfied).
>        
>        1. ❌ If the path is in _a different_ Git repo, the ideal behavior would be to show the correct Git status.
>        2. ❌ If the path is _not in any_ Git repo, the ideal behaviour would be to not even show the Git column.

I disagree about the ideal behaviour noted here, simply because that is diverging from `git` upstream: as long as `GIT_DIR` is set, that will be the current git directory. It overrides all path searches.

For instance, even if the file is in a different repo, if we try to `git add somepath`, git will respect `GIT_DIR`, and NOT the closer repo. So then why should `eza` try to diverge from git behaviour by trying to guess at the "real" git directory of that file? Shouldn't the output of `eza --git` match that of `git status`?

The user would need to unset the `GIT_DIR` for all other usage (case 2 mentioned above).

I agree that this is a bit counterintuitive, but trying to diverge from git is just going to cause more confusion.

**Comment by iamkroot** [comment]
@mentalisttraceur any thoughts?

**Comment by mentalisttraceur** [comment]
@iamkroot yeah, but got busy, sorry for not living up to the within-a-week estimate I gave initially.

**Comment by mentalisttraceur** [comment]
Here's a barely-edited copy-paste of what I jotted down the day that you brought this up:

Very initial thoughts:

I'd like to figure out how to make @iamkroot 's case work, ideally while preserving actually sensible behavior for things that are version-controlled by other Git repos or not version-controlled at all.

I suspect that just taking into account `GIT_WORK_TREE` for lookups against `GIT_DIR` will be sufficient. (If I remember right, I only tested `GIT_DIR` cases where `GIT_WORK_TREE` wouldn't have made a difference - in the absence of `GIT_WORK_TREE`, Git uses the current working directory, and I only tested `GIT_DIR` cases from within the base of a repo.)

The thing is, Git's `GIT_DIR` override-all-lookups only makes sense for `git` because it has a one-to-one mapping. A typical `git` invocation works with **exactly one** Git directory, so a single path in `GIT_DIR` overrides the **only** Git directory. But an `eza` invocation works with [0-∞) Git directories. One `GIT_DIR` to override all Git lookups becomes somewhere between nonsensical and wrong when asked to look at multiple Git repos at once (I had memories of writing this but I couldn't find it in my public comments... I must've cut it to try to stay concise.)

**Comment by iamkroot** [comment]
> One `GIT_DIR` to override all Git lookups becomes somewhere between nonsensical and wrong when asked to look at multiple Git repos at once

Could you provide a few examples when it is nonsensical? Specifically, when `GIT_DIR` is set, I expect _all_ of my `git` status values to be relative to that, regardless of the current dir. If you disagree with this, I would like to understand the rationale.

**Comment by mentalisttraceur** [comment]
@iamkroot I don't know your life+workflows enough to know what would feel sufficiently compellingly "real-life" for you, and I don't really think in examples, but...

Think about what the actual executable logic of Git status lookup is supposed to do if a user does

`GIT_DIR=~/foo eza . ../other-dir`

..bearing in mind that [per Git documentation](https://git-scm.com/book/en/v2/Git-Internals-Environment-Variables), setting `GIT_DIR` inherently makes things relative to the current directory:

> If `--git-dir` or `GIT_DIR` is specified but none of `--work-tree`, `GIT_WORK_TREE` or `core.worktree` is specified, the current working directory is regarded as the top level of your working tree.

Personally, I frequently run `ls` or `eza` on paths starting with `../`, `/`, or `~/` while inside a Git repo, including into other Git repos sometimes. And if my current repo happens to be a repo for which I need `GIT_DIR` set, then I would probably have it exported in the environment, and I expect `eza` to **do what `git` does** in those situations: **not** consider paths as tracked by `GIT_DIR` if they are outside its work tree.

And if Eza detects, just like Git would, that a file is not in the work tree managed by `GIT_DIR`, then I expect Eza to do its normal Git lookup for those paths, because to do otherwise is needlessly unhelpful and at worst actively misleading about status information in ways users would not expect.

Concretely: when I do `git status ../other-dir`, if I have `GIT_DIR` set but that other repo is not part of what git would consider the worktree of `GIT_DIR`, I don't expect Git to report zero changes - Git errors out saying the path is outside the repository. Similarly, if I do `eza -l --git ../other-dir` in the same situation, the *last* thing I expect is for it to show a git status column that has nothing to do with what the actual status over there is, because it silently did the lookup against `GIT_DIR` - and yet that was the incompatible behavior that Eza was doing before my code change.

And part of why I expect the behavior of a separate lookup for outside-the-repo paths is because Eza is in the business of listing multiple directories, possibly covered by multiple repos, and **in every other situation** Eza does not use a Gitdir for a path's status if that path does not belong to that Gitdir's worktree per Git's rules/behavior.

Concretely: if I am in a repo and I do `git status . ../other-repo`, that's an error because there can only be one repo consulted, but if I do `eza -l --git . ../other-repo`, that's a second directory listing with a second git repo consulted.

**Comment by mentalisttraceur** [comment]
Or to bring back your example, if I have this setup like you do:

```sh
export GIT_ENV="$HOME/.local/share/yadm/repo.git"
export GIT_WORK_TREE="$HOME"
```

it would be somewhere between wrong and non-sensical for `GIT_DIR` to be used for the status column of `eza --git --long /dev`, because that's outside of that worktree by Git's own logic, and Eza *was* previously lying in that case (it showed a git status column with no changes instead of no git status column, if I remember right).

**Comment by mentalisttraceur** [comment]
To be clear, **I fully agree** that what I implemented was wrong for some cases.

(That wasn't immediately clear to me when you first brought this up, and I'm sorry both for missing it and for being initially dismissive/disbelieving/misunderstanding.)

And I do intend to work on fixing this, hopefully very soon now that I fully understand the problem - I think `eza` should definitely use `GIT_DIR` for the status of any paths that fall inside its work tree (which includes making sure it respects the precedence of `GIT_WORK_TREE` -> `core.worktree` -> current working directory).

I'm just saying that we should still keep normal Git repo lookups **for paths outside `GIT_DIR`'s worktree** - I think that is useful+helpful, consistent with Git (about as much as `eza` being able to look at more than one repo within a single execution) in an intuitive way, and won't break any use-cases like yours once the above is fixed.

**Comment by iamkroot** [comment]
My only question is, what is stopping the user from unsetting the `GIT_DIR` env when they want to look at an unrelated directory? Why does it have to be set all the time?

**Comment by mentalisttraceur** [comment]
I finally found time to start looking at it again, but it looks like in the meantime, the behavior I added regressed in 90ee3b0b999b90441dbaf67d845a07a9bd94ad4e (the upgrade of `git2` from 0.17.2 to 0.18.0).

Maybe in the process it automatically started doing what @iamkroot wanted again?

In the meantime, I've basically stopped using `eza`, because in my life it has been superseded by other workflow improvements (*especially* w.r.t. `git`), so I am now mostly unaffected by `eza` (not) respecting Git's ceiling directories. If I ever have the intersection of time+need again, I'll come back with a patch that handles all of these cases. Until then, please proceed without me, however is best for people who are actually still using `eza --git`, even if that means just reverting my patch.

**Comment by cafkafk** [comment]
I don't expect to see progress on this anytime soon, if it reappears as an issue, we can reopen it
